[TypeScript docs] document WorkflowService advanced usage (#820)
temporalio/documentation: docs/typescript/clients.md

  Given a `workflowId`, Workflow Clients may also get a Handle to a running Workflow Execution or retrieve/wait for its result.
- **Workflow Handles** are bindings to specific Workflow Executions that expose more APIs for control.
  **We strongly recommend familiarising yourself with Workflow Handle APIs** as they are the main way you will signal, query, describe, cancel, terminate and await the result of running Workflow Executions.

> - Advanced users can also use the `WorkflowService` exposed by a Workflow Client to make **raw gRPC calls** (usually for introspection)

Reviewer: Add a period at the end of this sentence.

Using a Workflow Handle isn't necessary with `client.execute` by definition.

- **Don't forget to handle errors here** - if you call `result()` on a Workflow that prematurely ended for some reason, it will [throw an Error](https://typescript.temporal.io/api/classes/client.WorkflowFailedError) reflecting that reason.
  ```ts

>    const handle = client.getHandle(workflowId);
>    try {
>     	const result = await handle.result(); // block until the workflow completes, if you wish
>   	} catch (err) {
>   	  if (err instanceOf WorkflowFailedError) {
>   	     throw new Error('Temporal workflow failed: ' + workflowId), err)
>       } else {
>          throw new Error('error from Temporal workflow ' + workflowId), err)
>     }

Author: ```suggestion
   const handle = client.getHandle(workflowId);
   try {
    	const result = await handle.result(); // block until the workflow completes, if you wish
  	} catch (err) {
  	  if (err instanceOf WorkflowFailedError) {
  	     throw new Error('Temporal workflow failed: ' + workflowId, { cause: err})
      } else {
         throw new Error('error from Temporal workflow ' + workflowId, {cause: err})
    }
```

Using a Workflow Handle isn't necessary with `client.execute` by definition.

- **Don't forget to handle errors here** - if you call `result()` on a Workflow that prematurely ended for some reason, it will [throw an Error](https://typescript.temporal.io/api/classes/client.WorkflowFailedError) reflecting that reason.

>   `ts const handle = client.getHandle(workflowId); try { const result = await handle.result(); // block until the workflow completes, if you wish } catch (err) { if (err instanceOf WorkflowFailedError) { throw new Error('Temporal workflow failed: ' + workflowId), err) } else { throw new Error('error from Temporal workflow ' + workflowId), err) } } `

Author: ```suggestion
  ```ts
   const handle = client.getHandle(workflowId);
   try {
    	const result = await handle.result(); // block until the workflow completes, if you wish 
  	} catch (err) {
  	  if (err instanceOf WorkflowFailedError) {
  	     throw new Error('Temporal workflow failed: ' + workflowId), err)
      } else {
         throw new Error('error from Temporal workflow ' + workflowId), err)
    }
   } 
   ```
```

This Service is capable of making a wider range of introspection calls (as per [the API reference](https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#methods)):

```ts
import { Connection, WorkflowClient } from '@temporalio/client';

const connection = new Connection(); // to configure connection for production

// normal Workflow client
const client = new WorkflowClient(connection.service);
await client.start(/* etc */);

// equivalent in raw gRPC calls
// promise-based API
await connection.service.startWorkflowExecution(/* etc */);
// alternate Node callback-style API
connection.service.startWorkflowExecution(/* options */, callbackFn);
```

Using gRPC calls is often the only way to access some of the more advanced queries you can make from Temporal Server.
We highlight some queries of interest here:

- [describeTaskQueue](https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#describetaskqueue)
- [getWorkflowExecutionHistory](https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#getworkflowexecutionhistory)

> - [listWorkflowExecutions](https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#listworkflowexecutions)

Author: ```suggestion
- [listWorkflowExecutions](https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#listworkflowexecutions)

```ts
  const res = await connection.service.listWorkflowExecutions({ namespace: 'default'})
  console.table(res.executions)
```

Outputs:

```

┌─────────┬───────────────────────────────────────────────────────────────────────────────────────────────────┬──────────────────────────────────┬─────────────────────────────────────────────────┬─────────────────────────────────────────────────┬────────┬───────────────┬─────────────────────────────────────────────────┬─────────────────────┬──────────────────────────────────────────────┬────────────┬──────────────────────┐
│ (index) │                                             execution                                             │               type               │                    startTime                    │                    closeTime                    │ status │ historyLength │                  executionTime                  │        memo         │               searchAttributes               │ taskQueue  │ stateTransitionCount │
├─────────┼───────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────────────────┼─────────────────────────────────────────────────┼─────────────────────────────────────────────────┼────────┼───────────────┼─────────────────────────────────────────────────┼─────────────────────┼──────────────────────────────────────────────┼────────────┼──────────────────────┤
│    0    │ WorkflowExecution { workflowId: 'my-business-id', runId: '2798482a-46d8-4f1e-ab87-1ba3f7ddda1c' } │ WorkflowType { name: 'example' } │ Timestamp { seconds: [Long], nanos: 125158275 } │ Timestamp { seconds: [Long], nanos: 263021256 } │   2    │    [Long]     │ Timestamp { seconds: [Long], nanos: 125158275 } │ Memo { fields: {} } │ SearchAttributes { indexedFields: [Object] } │ 'tutorial' │        [Long]        │
│    1    │ WorkflowExecution { workflowId: 'my-business-id', runId: '76f1171b-7a73-46a3-ba66-b77bab0b73f8' } │ WorkflowType { name: 'example' } │ Timestamp { seconds: [Long], nanos: 841243925 } │ Timestamp { seconds: [Long], nanos: 935283589 } │   2    │    [Long]     │ Timestamp { seconds: [Long], nanos: 841243925 } │ Memo { fields: {} } │ SearchAttributes { indexedFields: [Object] } │ 'tutorial' │        [Long]        │
│    2    │ WorkflowExecution { workflowId: 'my-business-id', runId: '6d1197b7-41b8-47be-89b4-f1ef3446de1a' } │ WorkflowType { name: 'example' } │ Timestamp { seconds: [Long], nanos: 425778697 } │ Timestamp { seconds: [Long], nanos: 523022091 } │   2    │    [Long]     │ Timestamp { seconds: [Long], nanos: 425778697 } │ Memo { fields: {} } │ SearchAttributes { indexedFields: [Object] } │ 'tutorial' │        [Long]        │
└─────────┴───────────────────────────────────────────────────────────────────────────────────────────────────┴──────────────────────────────────┴─────────────────────────────────────────────────┴─────────────────────────────────────────────────┴────────┴───────────────┴─────────────────────────────────────────────────┴─────────────────────┴──────────────────────────────────────────────┴────────────┴──────────────────────┘
```

```

```

Again, see the [Workflow APIs documentation](/docs/typescript/workflows#external-workflows) for full details.

## Advanced: Making raw gRPC calls

Under the hood of a `WorkflowClient`, the `Connection` is actually powered by a `WorkflowService` driver that makes the raw gRPC calls to Temporal Server.
This Service is capable of making a wider range of introspection calls (as per [the API reference](https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#methods)):

> ```ts

Reviewer 2: I'd really rather stop putting snippets directly into the docs, we have snipsync, let's use it so we know the snippets work.

Author: yeah ok ok on it

We highlight some queries of interest here:

<details>
<summary><a class="font-mono" href="https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#getworkflowexecutionhistory">getWorkflowExecutionHistory</a>
</summary>

```ts
const res = await connection.service.getWorkflowExecutionHistory({
  execution: { workflowId: 'my-business-id' },
  namespace: 'default',
});
console.log(res.history);
```

Outputs:

```
{
  events: [
    HistoryEvent {
      eventId: [Long],
      eventTime: [Timestamp],
      eventType: 1,
      taskId: [Long],
      workflowExecutionStartedEventAttributes: [WorkflowExecutionStartedEventAttributes]
    },
    HistoryEvent {
      eventId: [Long],
      eventTime: [Timestamp],
      eventType: 5,
      taskId: [Long],
      workflowTaskScheduledEventAttributes: [WorkflowTaskScheduledEventAttributes]
    },
    HistoryEvent {
      eventId: [Long],
      eventTime: [Timestamp],
      eventType: 6,
      taskId: [Long],
      workflowTaskStartedEventAttributes: [WorkflowTaskStartedEventAttributes]
    }
  ]
}
```

</details>

<details>

> <summary><a class="font-mono" href="https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#listworkflowexecutions) (requires ElasticSearch">listWorkflowExecutions</a>

Reviewer 2: I don't think this requires elasticsearch if you use basic queries.

Author: i tried it, it does

Author: ```suggestion
<summary><a class="font-mono" href="https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#listworkflowexecutions) (requires ElasticSearch)">listWorkflowExecutions</a>
```

}
```

Again, see [Workflows in TypeScript](/docs/typescript/workflows#external-workflows) for full details.

> ## Advanced: Making raw gRPC calls

Reviewer 3: Any chance you would be willing to convert this to a "how to" page?

Reviewer 3: Go ahead and get this in as-is.

```

Again, see the [Workflow APIs documentation](/docs/typescript/workflows#external-workflows) for full details.

## Advanced: Making raw gRPC calls

Under the hood of a `WorkflowClient`, the `Connection` is actually powered by a `WorkflowService` driver that makes the raw gRPC calls to Temporal Server.

> This Service is capable of making a wider range of introspection calls (as per [the API reference](https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#methods)):

Reviewer 4: ```suggestion
This Service is capable of making a wider range of introspection calls. (For more information, see the Methods section of the [WorkflowService](https://typescript.temporal.io/api/classes/proto.temporal.api.workflowservice.v1.WorkflowService-1#methods)) API reference.
```
